File: yaf-gadomain/src/main/java/ar/edu/ungs/yamiko/ga/domain/impl/BasicGenome.java
package ar.edu.ungs.yamiko.ga.domain.impl;

import java.io.Serializable;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ar.edu.ungs.yamiko.ga.domain.Gene;
import ar.edu.ungs.yamiko.ga.domain.Genome;
import ar.edu.ungs.yamiko.ga.domain.Ribosome;
import ar.edu.ungs.yamiko.ga.toolkit.StaticHelper;

/**
 * Implementación de Genoma orientado a tiras de longitud estática de tipo T.
 * @author ricardo
 *
 */
public class BasicGenome<T> implements Genome<T>,Serializable{

	/**
	 * 
	 */
	private static final long serialVersionUID = -8422798873296370912L;
	private Map<String,List<Gene>> structure;
	private int size;
	private Map<Gene,Ribosome<T>> translators;

	public BasicGenome(String uniqueCrhomosomeName,List<Gene> genes, Map<Gene,Ribosome<T>> _translators) 
	{
		structure=new HashMap<String, List<Gene>>();
		structure.put(uniqueCrhomosomeName,genes);
		size=StaticHelper.calcGeneSize(structure.get(uniqueCrhomosomeName));
		translators=_translators;
	}
	
	public Map<String, List<Gene>> getStructure() {
		return structure;
		
		}
	
	public int size() {
		return size;
	}
		
	public Map<Gene,Ribosome<T>> getTranslators() {
			return translators;
		}
}

File: yaf-gadomain-scala/src/main/scala/ar/edu/ungs/yamiko/ga/domain/impl/BasicGenome.scala
package ar.edu.ungs.yamiko.ga.domain.impl

import ar.edu.ungs.yamiko.ga.domain.Gene
import ar.edu.ungs.yamiko.ga.domain.Genome
import ar.edu.ungs.yamiko.ga.domain.Ribosome


@SerialVersionUID(1159L)
class BasicGenome[T](uniqueCrhomosomeName:String, genes:List[Gene], translators:Map[Gene,Ribosome[T]]) extends Genome[T]{
  
  //Constructor
	val structure:Map[String, List[Gene]]=Map(uniqueCrhomosomeName -> genes)
	val _size=genes.map(_.size()).sum
  
  override def size():Int=_size
  override def getStructure()=structure;
  override def getTranslators() =translators
    
	override def toString = "BasicGenome [size=" + size + ", uniqueCrhomosomeName=" + uniqueCrhomosomeName + "]"
	
  def canEqual(a: Any) = a.isInstanceOf[BasicGenome[T]]

  override def equals(that: Any): Boolean =
    that match {
      case that: BasicGenome[T] => 
        that.canEqual(this) && this.hashCode == that.hashCode
      case _ => false
  }

  override def hashCode:Int = {
    val ourHash = if (uniqueCrhomosomeName == null) 0 else uniqueCrhomosomeName.hashCode+size.hashCode()    
    super.hashCode + ourHash
  }
  
}